main.cpp

```diff
@@ -0,0 +1,201 @@
+#include <iostream>
+#include <stdbool.h>
+#include <vector>
+#include <tuple>
+
+#include <climits>
+
+#include <cmath>
+#include <cassert>
+
+typedef int dtype; 
+
+class Square {
+private:
+    dtype num; 
+    //List of Possibilities
+
+public:
+    Square(): num(-1) {}
+    Square(dtype val): num(val) {}
+
+    bool is_solved() { return (num < 0) ? false : true; };
+    dtype& value() {
+        return num;
+    }
+    
+    //override = operator, so can assign value directly
+    void operator= (const dtype num) {
+        this->num = num;
+    }
+};
+
+class Grid {
+public:
+    class GridRow
+    {
+        private:
+            friend class Grid; // Allow Grid to Create & Access Internals
+            std::vector<Square> vec;
+            GridRow(const size_t size) : vec(size, Square()){}
+            GridRow(const std::vector<dtype> v) {
+                for(auto s: v) {
+                    vec.emplace_back(Square(s));
+                }
+            } 
+            
+        public:
+            Square& operator[](size_t idx) {return vec[idx];}
+            size_t size() {return vec.size();}
+            std::vector<dtype> values() {
+                std::vector<dtype> res;
+                for(auto sq: vec) {
+                    res.emplace_back(sq.value());
+                }
+                return res;
+            }
+    };
+    
+private:
+    size_t grid_size; //row & col 
+    size_t block_len; // sqrt(row)
+
+    std::vector<GridRow> grid;
+    
+    //checks is the vectors are possible in sudoku
+    //returns : bool - true or false if possible 
+    //   size_t - number of digits missing in row (only valid when bool is true)
+    std::pair<bool, size_t> check_if_vecs_possible(const 
+                                        std::vector<dtype>& vec) {
+        if(vec.size() != grid_size) {
+            return std::make_pair(false, 0 );
+        }
+        
+        //use a vector to check if value has been seen
+        std::vector<bool> seen(vec.size(), false);
+        size_t num_left = grid_size;
+        
+        for(auto val: vec) {
+            if(val < 0) {
+                continue;

```

Comment:
do you need num_left -- here as well?

Reply:
No.  It should be a better variable name, but num_left is the numbers that are currently not seen.